controllers/component_dependency_update_controller.go

```diff
@@ -0,0 +1,224 @@
+// +kubebuilder:rbac:groups=tekton.dev,resources=pipelineruns/finalizers,verbs=update
+// +kubebuilder:rbac:groups="",resources=events,verbs=create;patch
+func (r *ComponentDependencyUpdateReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	var cancel context.CancelFunc
+	ctx, cancel = context.WithTimeout(ctx, contextTimeout)
+	defer cancel()
+	log := ctrllog.FromContext(ctx).WithName("ComponentNudge").WithValues("namespace", req.NamespacedName.Namespace, "resource", req.Name)
+	ctx = ctrllog.IntoContext(ctx, log)
+
+	pipelineRun := &tektonapi.PipelineRun{}
+	err := r.Get(ctx, req.NamespacedName, pipelineRun)
+	if err != nil {
+		log.Error(err, "Failed to get pipelineRun")
+		if errors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	component, err := GetComponentFromPipelineRun(r.Client, ctx, pipelineRun)
+	if err != nil || component == nil {
+		log.Error(err, "failed to get component")
+		return ctrl.Result{}, err
+	}
+
+	if len(component.Spec.BuildNudgesRef) == 0 {
+		log.Info(fmt.Sprintf("component %s has no BuildNudgesRef set", component.Name))
+		return ctrl.Result{}, nil
+	}
+
+	if pipelineRun.IsDone() || pipelineRun.DeletionTimestamp != nil {
+		if controllerutil.ContainsFinalizer(pipelineRun, NudgeFinalizer) || pipelineRun.Annotations == nil || pipelineRun.Annotations[NudgeProcessedAnnotationName] == "" {
+			// Pipeline run is done and we have not cleared the finalizer yet
+			// We need to perform our nudge
+			return r.handleCompletedBuild(ctx, log, pipelineRun, component)
+		}
+	} else if !controllerutil.ContainsFinalizer(pipelineRun, NudgeFinalizer) {
+		// We add a finalizer to make sure we see the run before it is deleted
+		// As tekton results should aggressivly delete when pruning is enabled
+		controllerutil.AddFinalizer(pipelineRun, NudgeFinalizer)
+		err := r.Client.Update(ctx, pipelineRun)
+		if err != nil {
+			log.Error(err, "failed to add finalizer")
+			return ctrl.Result{}, err
+		}
+	}
+	return ctrl.Result{}, nil
+}
+
+// handleCompletedBuild will perform a 'nudge' updating dependent downstream components
+// this will involve creating a PR updating their references to our images to the newly produced image
+// for now it just produces a log message
+func (r *ComponentDependencyUpdateReconciler) handleCompletedBuild(ctx context.Context, log logr.Logger, pipelineRun *tektonapi.PipelineRun, updatedComponent *applicationapiv1alpha1.Component) (ctrl.Result, error) {
+
+	// Note that we do this eagerly, as we only really have one chance to do the nudge
+	// We don't want to get stuck in a nudge loop, or fail to remove the finalizer
+	if pipelineRun.Annotations == nil {
+		pipelineRun.Annotations[NudgeProcessedAnnotationName] = "true"
+	}
+	controllerutil.RemoveFinalizer(pipelineRun, NudgeFinalizer)
+	err := r.Client.Update(ctx, pipelineRun)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
```

Comment:
This may determine any failure nudge will be a permanent error and the retry will not depend on reconcile, it needs another PipelineRun triggered again. Is my understanding correct?

Reply:
See this comment: https://github.com/redhat-appstudio/build-service/pull/210#discussion_r1382294356

This story was just about the initial reconcile loop and logging a message.